rir/src/compiler/analysis/reference_count.h

```diff
@@ -8,29 +8,92 @@ namespace rir {
 namespace pir {
 
 struct AUses {
+
+    /*
+     * Lattice:
+     *
+     *                    Multiple
+     *                     /      \
+     *   AlreadyIncremented        Once
+     *                     \      /
+     *                       None
+     *
+     * Definitions:
+     *   reuse            :=   override if refcount is == 0
+     */
+
     enum Kind {
-        None = 0,
-        Once = 1,
-        Multiple = 2,
-        Destructive = 3,
+        // 1) Static refcount is safe
+        None,
+        Once,               // At least one use, which reuses input
+        AlreadyIncremented, // Refcount is known to be 1 (ie. used by stvar)
+
+        // 2) Need an ensureNamed
+        Multiple, // At least two uses, at least one (but the last) is reusing
     };
 
-    std::unordered_map<Instruction*, Kind> uses;
+    static Kind merge(Kind a, Kind b) {
+        if (a == b)
+            return a;
+
+        Kind bigger = a > b ? a : b;
+        Kind smaller = a < b ? a : b;
+        if (smaller == None)
+            return bigger;
+
+        if (bigger == Multiple || bigger == AlreadyIncremented)
+            return Multiple;
+
+        assert(false);
+        return Multiple;
+    };
+
+    bool overflow = false;
+    std::map<Instruction*, Kind> uses;
```

Comment:
If this needs to be ordered, a comment saying so would be nice, otherwise someone might change it to `unordered_map`.